### ISSUE TITLE
Terraform show -json fails in 3.71.0 onwards

### ISSUE THREAD

**Issue by KyleKotowick** [hashicorp/terraform-provider-aws#22854]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform v1.1.4
on windows_amd64
+ provider registry.terraform.io/hashicorp/aws v3.71.0

### Affected Resource(s)

* `terraform show -json`

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

I've tried to reproduce this, but I can't find which resources are causing it. I've reviewed the state file for everything with a field named `kms_key_arn` (the only thing is an `aws_lambda_function` resource), and when I try to use that resource in a clean config, it works fine.

### Panic Output

`Failed to marshal state to json: unsupported attribute "kms_key_arn"`

### Expected Behavior

It should output the JSON-encoded state.

### Actual Behavior

It fails.

### Steps to Reproduce

1. `terraform show -json`

### Important Factoids

This works fine with the AWS provider v3.70.0. As soon as I upgrade to 3.71.0, it throws this error, even though nothing in my config (other than the provider version) or state file changes. If I downgrade again, it works again.

**Comment by justinretzolk** [comment]
Hey @KyleKotowick 👋 Thank you for taking the time to raise this. Are you using any `aws_dynamodb_table` resources by chance? The error and timing (re: `v3.71.0`) seems eerily similar to https://github.com/hashicorp/terraform-provider-aws/issues/22534.

**Comment by KyleKotowick** [comment]
@justinretzolk Yep, it's definitely the DynamoDB table. I just did a `terraform show` and got this:
```
# module.mymodule.aws_dynamodb_table.sometable:
resource "aws_dynamodb_table" "sometable" {# 
```
And that's it for that resource, it just stopped outputting it after the opening bracket (note the trailing `#`, which is indicates the start of the next resource).

**Comment by KyleKotowick** [comment]
I tried to create a minimal example for this, but even with copying the exact DynamoDB table configuration to a fresh config, it works fine. Even tried sticking it inside a module like my main config that's throwing the error. It works fine though, I can't seem to reproduce it separately.

**Comment by KyleKotowick** [comment]
This is still an issue in the latest version (4.3.0).

**Comment by KyleKotowick** [comment]
OK, I think I found a workaround/fix.

1. Remove all DynamoDB tables from the state file (`terraform state rm "aws_dynamodb_table.foo"`)
2. Upgrade the provider if necessary (I upgraded to `4.3.0`, `terraform init -upgrade`)
3. Refresh the state (`terraform refresh`)
4. Import the Dynamo DB tables again (`terraform import "aws_dynamodb_table.foo" "my-table-name"`)

Now I'm not getting any errors when I run `terraform show -json`. My guess is that somewhere around version `3.71.0`, there was a change that affected the state file.

**Comment by justinretzolk** [comment]
Hey @KyleKotowick 👋 Thank you for following up with additional information in case anyone else runs into this in the future! Given that this seems to be tied to a specific provider version, do you feel comfortable with me marking this issue as closed?

**Comment by KyleKotowick** [comment]
@justinretzolk It's not an issue with a specific version; it's an issue for anyone who's upgrading from `< 3.71.0` to `>= 3.71.0`. i.e. anyone who had a version older than `3.71.0` and upgrades to a version newer than (or equal to) `3.71.0` who has a DynamoDB table may encounter this.

I've personally encountered this exact issue in 3+ configurations that I've upgraded recently, and they all had different initial versions (3.40, 3.68, 3.70) and were upgraded to different new versions (whatever was most recent at the time of the upgrade: 3.74, 4.0, 4.3).

**Comment by justinretzolk** [comment]
Hey @KyleKotowick -- thank you for that additional context. I've added a couple of additional labels, and will leave this open so we can take a look at it 👍

**Comment by philm** [comment]
I ran into the same error but for a different attribute:
```Failed to marshal state to json: unsupported attribute "ipv4_ipam_pool_id"```

_aws_dynamodb_table_ is not part of my project.

This error appears when ```terraform show -json``` is applied. I'm using Terraform Cloud configured for "remote" execution but not connected to a VCS provider. When ```terraform apply``` is executed, Terraform Cloud implicitly calls ```terraform show -json``` and errors out.

The error went away after running ```terraform apply -refresh-only``` so seems like state was out of sync.

Using Terraform v1.1.9 and AWS provider 4.10.0.

**Comment by samlarsen1** [comment]
We've seen this on at least 3 Terraform Enterprise workspace upgrades over the last 9 months.  Plans are successful and the post-plan stage `terraform show -json /terraform/terraform.tfplan` fails with "Failed to marshal plan to json: error in marshalResourceDrift ... unsupported attribute role_last_used" 
We're currently upgrading  `4.60.0`

**Comment by kishaningithub** [comment]
## TLDR version 

I also ran into the same issue, I fixed it by running the below targeted terraform refresh command 2 times (yeah you read it right 2 times)

```bash
terraform refresh -target=aws_dynamodb_table.table1 -target=aws_dynamodb_table.table2
```

## Additional Context
I was using terraform version `0.13.7` and aws provider version `4.9.0`. To inspect, I downloaded the state file from the S3 bucket where it was stored and found out that `kms_key_arn` was present was within the `ttl` attribute even though no such attribute is documented in the provider documentation for `4.9.0` version

https://registry.terraform.io/providers/-/aws/4.9.0/docs/resources/dynamodb_table#ttl


**Excerpt from the state file**

```json
{
  "mode": "managed",
  "type": "aws_dynamodb_table",
  "name": "table1",
  "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
  "instances": [
    {
      "schema_version": 1,
      "attributes": {
        ...
        "ttl": [
          {
            "attribute_name": "",
            "enabled": false,
            "kms_key_arn": "" // This should not exist here as per provider documentation
          }
        ],
        ...
      }
    }
    ...
  ]
}
```

Once i ran the targeted refresh (command above) then the `kms_key_arn` attribute vanished 😊 But again the `terraform show -json` command failed with the error

```
Failed to marshal state to json: unsupported attribute "deletion_protection_enabled"
```

when i inspected the state file again i found this `deletion_protection_enabled` which also should not have come as per the provider documentation

https://registry.terraform.io/providers/-/aws/4.9.0/docs/resources/dynamodb_table

```json
{
  "mode": "managed",
  "type": "aws_dynamodb_table",
  "name": "table1",
  "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
  "instances": [
    {
      "schema_version": 1,
      "attributes": {
        "deletion_protection_enabled": "" // This should not exist here as per provider documentation
        ...
      }
    }
    ...
  ]
}
```

So i ran the same targeted refresh command once again (they say keep trying until you succeed 😄) and it removed that attribute as well and `terraform show -json` command started working as expected 🎉
